### PR TITLE
Perlcritic tests

### PR DESCRIPTION
**Pull request by ferki**
This is still heavily work in progress, but overall I believe there are not so complicated fixes for all the violations at the current severity level, so I try to include them in this PR. If things get complicated or we decide the changes are too risky, we can still just mark the known violations as known ones for documentation purposes.

I preferred writing `## no critic ...` style comments to make sure we don't silently introduce further violations of a given policy (which is possible if we would just disable the policy completely in `.perlcriticrc`).

Depending on the decisions when the full picture is already known, I'll probably rewrite the commit messages before merging to make the history easier to understand.

### CODE REVIEW

**Comment by ferki** [comment]
Since it's taking too long, I've decided to limit the scope of this PR to just simply enable the perlcritic tests, and mark known violations as accepted. A followup PR is in progress with all the fixes (or that can be done in smaller chunks too, one PR per violated policies).

File: lib/Rex/Commands.pm [hunk 1]
```diff
@@ -1059,7 +1059,7 @@ sub needs {
   if ($self) {
     my $suffix = $self;
     $suffix =~ s/::/:/g;
-    @args = map { $_ = "$suffix:$_" } @args;
+    @args = map { $_ = "$suffix:$_" } @args; ## no critic ProhibitMutatingListFunctions
```
Comment (ehuelsmann) [inline lib/Rex/Commands.pm]: The code as it is there now, is equivalent to:

```perl
@args = map { "$suffix:$_" } @args;
```

due to the fact that the return value of `map` is assigned to the variable being iterated over. I think it's better, even for the scope of this PR not to suppress but to fix this instance.

File: lib/Rex/Commands/Run.pm [hunk 2]
```diff
@@ -255,7 +255,7 @@ sub run {
 
     if ( $args && ref($args) eq "ARRAY" ) {
       my $quoter = Net::OpenSSH::ShellQuoter->quoter( $exec->shell->name );
-      $cmd = "$cmd " . join( " ", map { $_ = $quoter->quote($_) } @{$args} );
+      $cmd = "$cmd " . join( " ", map { $_ = $quoter->quote($_) } @{$args} ); ## no critic ProhibitMutatingListFunctions
```
Comment (ehuelsmann) [inline lib/Rex/Commands/Run.pm]: `@args` isn't used in the remainder of the function, which leads to a similar argument as above:

The code as it is there now, is equivalent to:

```perl
@args = map { "$suffix:$_" } @args;
```

due to the fact that the mutated value of `@args` isn't used. I think it's better, even for the scope of this PR not to suppress but to fix this instance.

File: lib/Rex/Helper/Run.pm [hunk 3]
```diff
@@ -146,7 +146,7 @@ sub i_exec {
   my $exec   = Rex::Interface::Exec->create;
   my $quoter = Net::OpenSSH::ShellQuoter->quoter( $exec->shell->name );
 
-  my $_cmd_str = "$cmd " . join( " ", map { $_ = $quoter->quote($_) } @args );
+  my $_cmd_str = "$cmd " . join( " ", map { $_ = $quoter->quote($_) } @args ); ## no critic ProhibitMutatingListFunctions
```
Comment (ehuelsmann) [inline lib/Rex/Helper/Run.pm]: `@args` isn't used in the remainder of the function, which leads to a similar argument as above:

The code as it is there now, is equivalent to:

```perl
@args = map { "$suffix:$_" } @args;
```

due to the fact that the mutated value of `@args` isn't used. I think it's better, even for the scope of this PR not to suppress but to fix this instance.

File: lib/Rex/Helper/Run.pm [hunk 4]
```diff
@@ -157,7 +157,7 @@ sub i_exec_nohup {
...
```
Comment (ehuelsmann) [inline lib/Rex/Helper/Run.pm]: `@args` isn't used in the remainder of the function, which leads to a similar argument as above:

The code as it is there now, is equivalent to:

```perl
@args = map { "$suffix:$_" } @args;
```

due to the fact that the mutated value of `@args` isn't used. I think it's better, even for the scope of this PR not to suppress but to fix this instance.

**Comment by ferki** [comment]
Most of the fixes are already done and ready to be merged after this one from the [perlcritic-fix](https://github.com/RexOps/Rex/commits/perlcritic-fix) branch, including for the ones you highlighted.

If we really do want to mix marking and fixing violations here, perhaps it's best to draw the line at the "obvious" fixes, and include these three right away:
 - 823fad300873a2cb60ed03874a2997c92cb77170 Remove unnecessary subroutine prototype
 - da3f5537277204a18f03502ec2cd91a4d43cc71c Fix critic for ProhibitMutatingListFunctions
 - 91bc3336821bf28037d0cc6a58020d57ca715a06 Fix critic for ProhibitSleepViaSelect

@ehuelsmann: if you agree with those three commits, I can update this PR with them (fix instead of mark), leaving the rest for another PR(s).

**Comment by ehuelsmann** [comment]
Completely agreed with the three commits!

**Pull request merged**